src/ApiCore/ResourceTemplate/RelativeResourceTemplate.php

```diff
@@ -0,0 +1,296 @@
+        // match by examining the difference in count between the template segments and the
+        // path pieces.
+
+        if (count($pathPieces) < count($flattenedKeySegmentTuples)) {
+            // Each segment in $flattenedKeyedSegments must consume at least one
+            // segment in $pathSegments, so matching must fail.
+            throw $this->matchException($path);
+        }
+
+        $doubleWildcardPieceCount = count($pathPieces) - count($flattenedKeySegmentTuples) + 1;
+
+        $bindings = [];
+        $pathPiecesIndex = 0;
+        /** @var Segment $segment */
+        foreach ($flattenedKeySegmentTuples as list($segmentKey, $segment)) {
+            // So long as there are tuples remaining, there will also be $pathPieces
+            assert($pathPiecesIndex < count($pathPieces));
+
+            // In our flattened list of segments, we should never encounter a variable segment
+            assert($segment->getSegmentType() !== Segment::VARIABLE_SEGMENT);
+
+            if ($segment->getSegmentType() == Segment::DOUBLE_WILDCARD_SEGMENT) {
+                $pathPiecesForSegment = array_slice($pathPieces, $pathPiecesIndex, $doubleWildcardPieceCount);
+                $pathPiece = implode('/', $pathPiecesForSegment);
+                $pathPiecesIndex += $doubleWildcardPieceCount;
+            } else {
+                $pathPiece = $pathPieces[$pathPiecesIndex++];
+            }
+            if (!$segment->matches($pathPiece)) {
+                throw $this->matchException($path);
+            }
+
+            // If we have a valid key, add our $pathPiece to the $bindings array. Note that there
+            // may be multiple copies of the same $segmentKey. This is because a flattened variable
+            // segment can match multiple pieces from the path. We can add these to an array and
+            // collapse them all once the bindings are complete.
+            if (isset($segmentKey)) {
+                $bindings += [$segmentKey => []];
+                $bindings[$segmentKey][] = $pathPiece;
+            }
+        }
+
+        // It is possible that we have left over path pieces, which can occur if our template does
+        // not have a double wildcard. In that case, the match should fail.
+        if ($pathPiecesIndex !== count($pathPieces)) {
+            throw $this->matchException($path);
+        }
+
+        // Collapse the bindings from lists into strings
+        $collapsedBindings = [];
+        foreach ($bindings as $key => $boundPieces) {
+            $collapsedBindings[$key] = implode('/', $boundPieces);
+        }
+        return $collapsedBindings;
+    }
+
+    private function matchException($path)
+    {
+        return new ValidationException("Could not match path '$path' to template '$this'");
+    }
+
+    /**
+     * @param Segment[] $segments
+     * @return array[] A list of [string, Segment] tuples
+     */
+    private static function buildKeySegmentTuples($segments)
```

Comment:
We could use an array typehint here

Reply:
Fixed in https://github.com/googleapis/gax-php/pull/209